# 목차    
- [4. 예외](4.-예외)

<BR>

# **4. 예외**

## **4.1 사라진 SQLException**

3장에서 만든 deleteAll() 함수를 보면, jdbcTemplate을 사용하기 전과 후의 코드가 다른 부분이 있다.

```java
// Before
public void deleteAll() throws SQLException {
    this.jdbcContext.executeSql("delete from users");
}

// After
public void deleteAll() {
    this.jdbcTemplate.update("delete from users");
}
```

SQLException이 어디로 사라졌는지 알아보도록 하자.

<BR>

## **4.1.1 초난감 예외처리**
사라진 SQL Exception을 찾기 이전에, 다양한 초난감 예외처리 사례들을 보도록 하자.    

<BR>

### **예외 블랙홀**    
예외를 잡았으나 아무것도 하지 않고 정상 코드의 흐름으로 이어가는 경우.    
의도한게 아니라면 **절대 금지**
```java
try {
    ...
} catch(SQLException e) {
    // Nothing
}
```
<BR>

### **에러 출력형**
아래와 같은 형태의 코드를 왜 작성하면 안될까?     
=> 단순 출력을 통해 문제를 해결할 수가 없음.

println은 CPU 리소스를 점유하고, 디스크 I/O 동기화 처리가 되기 때문에 전체적인 시스템의 성능이 저하되는 이슈가 있음.    
실제 배포되는 서버에 남아있다면 매우 치명적.    

printStackTrace 또한 내부적으로 System.err을 잡고 있기 때문에 지양하는 형태.    
또한 Stack Tracing 정보를 노출하는 것 또한 보안상 지양하는 것이 좋음.

@Slf4j 같은 로깅 라이브러리를 이용하자.

```java
try {
    ...
} catch(SQLException e) {
    System.out.println(e);
}

// or
try {
    ...
} catch(SQLException e) {
    e.printStackTrace();
}
```
<BR>

### **무의미하고 무책임한 throws**
만약 method3에서 예외가 실제 발생했다면?    
method1을 호출한 위치에서 이에 대한 정보를 받고 처리해야하는 형태.    
또한 최상위 Exception Class를 Throw 함으로써 어떤 갈래의 에러인지도 알 수 없음.
```java
public void method1() throws Exception {
    method2();
    ...
}
public void method2() throws Exception {
    method3();
    ...
}
public void method3() throws Exception {
    ...
}
```

<BR>

## **4.1.2 예외의 종류의 특징**
JAVA에서 Throw를 통해 발생시킬 수 있는 예외는 크게 3가지가 있다.
<BR>

### **Error**
- java.lang.Error의 서브클래스들
- 시스템에 비정상적인 상황이 발생했을 때 사용
- 주로 JVM에서 발생하므로 Application Layer에서 처리하지 않아도 됨 


### **Exception & Checked Exception**
- java.lang.Exception의 서브클래스들
- 개발자가 만든 코드의 작업 중 예외상황이 발생했을 경우에 사용
- Exception Class는 Checked Exception / Unchecked Exception으로 나뉨 
- CheckedException의 경우 미리 처리하지 않으면 컴파일 에러 발생

### **RuntimeException & Unchecked Exception**
- java.langRuntimeException의 서브클래스들
- 명시적 예외처리를 강제하지 않아서 Unchecked Exception으로 불림
- 프로그램에 오류가 있을 때 발생하도록 의도된 것들
- catch나 throw를 사용하지 않아도 됨
- NullPointerException, IllegalArgumentException, etc..

<BR>

## **4.1.3 예외처리 방법**
다양한 예외처리 방법에 대해 알아보자.    

<BR>

### **예외 복구**
예외 상황을 파악하고 문제를 해결해서 정상 상황으로 되돌리는 것     
ex) DB Connection Fail -> **Retry**

```java
int maxRetry = MAX_RETRY;
while(maxRetry-- > 0) {
    try {
        ...
        return;
    }
    catch(SomeException e) {
        log.error("blah");
        wait(...)
    }
    finally {
        // close
    }
}
throw new RetryFailedException();
```

<BR>

### **예외 회피**
두가지 방법이 있다.
- throw를 선언해서 예외 발생시 자동으로 호출한쪽으로 예외 전달
- catch로 예외를 잡은 후 rethrow

방법도 중요하지만 명확하게 예외를 처리할 layer를 정하는 것 또한 중요하다.
```java
public void add() throws SQLException {
    try {
        // JDBC
    }
    catch(SQLException e) {
        log.error(...);
        throw e;
    }
}
```

<BR>

### **예외 전환**
아래와 같은 경우에 사용한다. 
- 내부에서 발생한 예외를 그대로 던지는 것이 그 상황에 대한 적절한 의미를 부여해주지 못하는 경우    

```java
public void add(User user) throws DuplicateUserIdException, SQLException {
    try {
        // JDBC
    } catch(SQLException e) {
        // error code == Duplicate Entry (1062) 인 경우
        if (e.getErrorCode() == MysqlErrorNumbers.ER_DUP_ENTRY){
            throw DuplicateUserIdException();
        } else {
            throw e; // SQLException
        }
    }
}
```

아래와 같이 선언하면 근본 원인이 되는 예외 또한 알 수 있다.
```java
catch(SQLException e) {
    ...
    throw DuplicateUserIdException(e);
}

// or

catch(SQLException e) {
    ...
    throw DuplicateUserIdException().initCause(e);
}
```

<BR>

- 예외를 처리하기 쉽고 단순하게 만들기 위해 포장하는 경우     
```java
try {
    OrderHome orderHome = EJBHomeFactory.getInstance().getOrderHome();
    Order order = orderHome.findByPrimaryKey(id);
} catch (NamingException ne) {
    throw new EJBException(ne);
} catch (SQLException se) {
    throw new EJBException(se);
} catch (RemoteException re) {
    throw new EJBException(re);
}
```

이런 예외와 같이 예외를 잡아도 복구할 방법이 없는 경우(시스템적인 예외)는 RuntimeException으로 만들어서 전달하는 방향으로 작성하는 것이 좋다.    

<BR>

## **4.1.4 예외처리 전략**
위의 케이스를 토대로 일관된 예외처리 전략을 보도록 하자.

<BR>

### **런타임 예외의 보편화**
자바 초기부터 있었던 JDK의 API와 달리 최근에 등장하는 표준 스펙 또는 오픈소스 프레임워크에서는 API가 발생시키는 예외를 CheckedException 대신 UncheckedException으로 정의하는 것을 일반화 하고 있다.    
예전에는 복구할 가능성이 조금이라도 있다면 CheckedException으로 정의했는데, 지금은 항상 복구할 수 있는 예외가 아니라면 UncheckedException으로 만드는 경향이 있다.    
또한 UncheckedException 이더라도 얼마든지 catch()를 통해 잡아서 복구하거나 처리할 수 있다.    
하지만 대개는 그렇지 않고, RuntimeException으로 포장해서 던져야 하는 경우가 대다수이기 때문에 API 차원에서 RuntimeException을 던지도록 만드는 것이다.    


<BR>

### **add() 메소드의 예외처리**
앞서 본 add() 함수의 예외처리 전략을 살펴보면, DuplicatedUserIdException과 SQLException 두 가지의 CheckedException을 던지게 되어있다. 

DuplicatedUserIdException의 경우 복구 가능한 예외이지만, SQLException의 경우 대부분 그렇지 않다.     
그렇기에 그냥 RuntimeException으로 포장해서 던져주어 다른 메소드가 신경쓰지 않도록 하는 편이 낫다.

```java
// exception class
public class DuplicateUserIdException extends RuntimeException {
    public DuplicateUserIdException(Throwable cause) {
        super(cause);
    }
}

// method
public void add() throws DuplicateUserIdException {
    try {
        // JDBC
    } catch(SQLException e) {
        // error code == Duplicate Entry (1062) 인 경우
        if (e.getErrorCode() == MysqlErrorNumbers.ER_DUP_ENTRY){
            throw DuplicateUserIdException();
        } else {
            throw RuntimeException(e); // SQLException 포장
        }
    }
}
```

위 과정을 통해 SQLException을 처리하기 위한 throws 선언도 필요 없어졌고, DuplicateUserIdException또한 이용 가능하게 되었다.    

<BR>

### **애플리케이션 예외**

애플리케이션 자체 로직에 의해 의도적으로 발생시키고 반드시 catch해서 무언가 조치를 취하도록 요구하는 예외를 말한다.

ex) 사용자가 출금을 수행할 때, 잔고의 계좌보다 더 큰 금액을 출금하고자 하는 경우    

위와 같은 경우 2가지 방법으로 예외를 처리할 수 있다.

1. 정상 출금 처리와 비정상 출금처리에 따른 return 값 분기
2. 잔고 부족과 같은 예외 상황에서 비즈니스적인 의미를 띤 예외 제공

일반적으로 2번 방법을 택한다.

```java
try {
    BigDecimal balance = account.withdraw(amount);
    // 정상 프로세스
} catch(InsufficientBalanceException e) { // CheckedException
    // InsufficientBalanceException에 담긴 인출 가능한 잔고 금액 정보를 가져옴
    BigDecimal availFunds = e.getAvailFunds();

    ...
 
    // 1. 현재 인출 가능한 최대 금액 제공
    // 2. 예외 발생 메시지 제공 등
    // 다양한 방법으로 대응
}
```

<BR>

## **4.1.5 SQLException은 어떻게 됐나?**
JDBCTemplate은 예외처리 전략을 따르고 있다.    
템플릿 내부에서 발생하는 모든 SQLException을 RuntimeException인 DataAccessException으로 포장해서 던져주고 있다.    
따라서 DAO Method에서 SQLException이 모두 사라진 것이다.
```java
public int update(final String sql) throws DataAccessException { ... }
```

<BR>

# **4.2 예외 전환**

## **4.2.1 JDBC의 한계**
JDBC API는 JAVA를 이용해 DB에 Access하는 방법을 추상화된 API의 형태로 제공하고, 각 DB 업계는 이를 토대로 드라이버를 제공한다.    
추상화된 정보를 토대로 작성한 코드이기에 데이터를 액세스하는 코드를 작성하는것에는 문제가 없다.    
하지만, 그렇다고 해서 DB를 자유롭게 전환하여 이용할 수 있다는 이야기는 아니다.    

두 가지 이유가 있다.    

<BR>

### **비표준 SQL**
DB 벤더 마다 문법이 다른 부분에서 나타나는 부분이다.    
대표적으로 limit. *(Oracle에서는 사용 불가)*

<BR>

### **호환성 없는 SQLException**
JDBC API를 보면 데이터 처리 중에 발생하는 다양한 예외를 SQLException 한 가지만 던지도록 설계되어 있다.    
DB 벤더마다 정의한 고유한 에러코드가 다르기 때문이다.

스프링에서 위 문제를 해결하는 방법을 보도록 하자.

<BR>

## **4.2.2 DB 에러 코드 매핑을 통한 전환**    
SQL 상태코드는 JDBC 드라이버를 제작할 때 들어가므로 벤더마다 제각각 이지만, DB 전용 에러코드는 DB에서 직접 제공해주는 것이기 때문에 버전이 올라가더라도 어느정도의 일관성을 확보할 수 있다.

이를 기반으로 DB 에러 코드와 예외 클래스와 매핑시켜 관리하는 형태이다.
JDBC Template에서 던지는 예외는 모두 DataAccessException의 서브클래스 이므로 아래와 같은 메타정보를 토대로 적절한 예외 클래스를 선택하게 할 수 있다.
```xml
<bean id="Oracle" class="org.springframework.jdbc.support.SQLErrorCodes">
    <property name="badSqlGrammerCodes">
        <value>900, 903 ... </value>
    </property>

    ...
```

## **4.2.3 DAO 인터페이스와 DataAccessException 계층구조**

우리가 DAO를 사용하는 이유는, 다양한 DB Access 기술에 대해 동일한 추상화된 행동을 선택하기 위함이다.    
그런데, 모든 Data Access 기술의 API에서 다음과 같이 선언할 수 있을까?
```java
public interface UserDao() {
    public void add(User user);
    ...
}
```

당장 떠오르는 어려움은 DB 벤더 마다 예외를 다르게 던지는 경우가 있을 것이다.
```java
public void add(User user) throws PersistentException // JPA
public void add(User user) throws HibernateException  // Hibernate
public void add(User user) throws JdoException        // JDO
...
```

물론 JDBC API를 제외하고는 모두 RuntimeException의 형태로 포장되어 예외를 던지기 때문에,     
실제로는 JDBC API의 SQLException만 RuntimeException으로 포장하면 해결된다. 

하지만, 각각의 RuntimeException에 따라 비즈니스 로직에서 의미있게 처리해야 하는 경우도 존재하는데, (중복 키 에러 등) 이 또한 결국 클라이언트에서 처리해야할 문제가 된다.

<BR>

### **데이터 액세스 예외 추상화와 DataAccessException 계층구조**
그래서 스프링은 DataAccessException을 통해 DB 관련 예외를 추상화 하고 있다.    

ex) Data Access를 부정확하게 사용한 경우    
예외는 InvalidDataAccessResourceUsageException이 던져진다.    
이를 구체적으로 세분화 하면 아래와 같다.
> JDBC : BadSqlGrammerException    
Hibernate : HibernateQueryException    

<BR>

## **4.2.4 기술에 독립적인 UserDao 만들기**
위 내용을 기반으로 그냥 Interface를 만들면 된다.    
DB 벤더에 맞게 로직만 별도 구성하면 된다.

```java
public interface UserDao {
    void add(User user);
    ...
}
```
```java
public class UserDaoJdbc {
    ...
}
public class UserDaoJpa {
    ...
}
public class UserDaoHibernate {
    ...
}
```

<BR>

### **테스트 보완**
Dao 테스트를 수행하는 경우, 엣지 케이스를 테스트할 때는 추상화된 예외를 통해 해결하는 방법 밖에 없을까?    
즉, DB 벤더마다 다르게 뱉어내는 RuntimeException을 케어할 순 없을까?

두 가지 형태로 해결할 수 있다.

- 직접 예외를 정의해두고, 각 Dao의 add() method에서 정의한 예외를 throw 하도록 설계
- SQLExceptionTranslator를 구현한 클래스를 통해 에러 코드를 전환

<BR>

## **4.3 정리**
- 예외 전환을 통해 의미 있는 예외를 전달하자. 